core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala

```diff
@@ -32,22 +32,33 @@ private[kafka] abstract class SingleSourceLogic[K, V, Msg](
     val partitionAssignedCB = getAsyncCallback[Set[TopicPartition]] { assignedTps =>
       tps ++= assignedTps
       log.log(partitionLogLevel, "Assigned partitions: {}. All partitions: {}", assignedTps, tps)
-      subscription.rebalanceListener.foreach {
-        _.tell(TopicPartitionsAssigned(subscription, assignedTps), sourceActor.ref)
-      }
       requestMessages()
     }
 
     val partitionRevokedCB = getAsyncCallback[Set[TopicPartition]] { revokedTps =>
       tps --= revokedTps
       log.log(partitionLogLevel, "Revoked partitions: {}. All partitions: {}", revokedTps, tps)
-      subscription.rebalanceListener.foreach {
-        _.tell(TopicPartitionsRevoked(subscription, revokedTps), sourceActor.ref)
-      }
     }
 
     def rebalanceListener: KafkaConsumerActor.ListenerCallbacks =
-      KafkaConsumerActor.ListenerCallbacks(partitionAssignedCB.invoke, partitionRevokedCB.invoke)
+      KafkaConsumerActor.ListenerCallbacks(
+        assignedTps => {
+          subscription.rebalanceListener.foreach {
+            _.tell(TopicPartitionsAssigned(subscription, assignedTps), sourceActor.ref)
+          }
```

Comment:
I am not sure if moving `TopicPartitionsAssigned` messages sends from the callback to here will make big of a difference.

`partitionAssignedCB.invoke` is also an actor message send to the stages ActorRef. Therefore messages for both callback calls will come in order and eventually `TopicPartitionsAssigned` and `TopicPartitionsRevoked` will also be sent in order.

Am I missing something?

Reply:
The `KafkaConsumerActor.WrappedAutoPausedListener` is mounted directly in the Kafka consumer and called on a Kafka consumer thread. It in turn calls `ListenerCallbacks.onAssign` and `ListenerCallbacks.onRevoke` directly without any async involved.

Reply:
Yes, but the implementation of `onAssign` and `onRevoke` sends actor messages. And before the change the actor messages were sent to the graph stage (async callback invocation) and after the change the message send is directly to the rebalance listener.

That is one less hop, but I do not see how this changes the semantics.

Reply:
Ok, what would be needed to solve #539 are synchronous callbacks.
That would require a new API.